core/capabilities/remote/registration/client.go

```diff
@@ -0,0 +1,138 @@
+}
+
+// Client is a shim for remote capabilities that support registration to a workflow.  It polls the Server to ensure
+// the registration stays live.  In the current implementation the Server shim will unregister any workflow that has
+// not been re-registered within the registrationExpiry interval.
+type Client struct {
+	services.StateMachine
+	lggr                logger.Logger
+	registrationMethod  string
+	registrationRefresh time.Duration
+	capInfo             commoncap.CapabilityInfo
+	capDonInfo          commoncap.DON
+	localDonInfo        commoncap.DON
+	dispatcher          registerDispatcher
+	registeredWorkflows map[string]*clientRegistration
+	mu                  sync.RWMutex
+	stopCh              services.StopChan
+	wg                  sync.WaitGroup
+}
+
+func NewClient(lggr logger.Logger, registrationMethod string, registrationRefresh time.Duration, capInfo commoncap.CapabilityInfo, capDonInfo commoncap.DON,
+	localDonInfo commoncap.DON, dispatcher registerDispatcher, registryType string) *Client {
+	return &Client{
+		lggr:                lggr.Named(registryType),
+		registrationMethod:  registrationMethod,
+		registrationRefresh: registrationRefresh,
+		capInfo:             capInfo,
+		capDonInfo:          capDonInfo,
+		localDonInfo:        localDonInfo,
+		dispatcher:          dispatcher,
+		registeredWorkflows: make(map[string]*clientRegistration),
+		stopCh:              make(services.StopChan),
+	}
+}
+
+func (r *Client) Start(_ context.Context) error {
+	return r.StartOnce(r.lggr.Name(), func() error {
+		r.wg.Add(1)
+		go func() {
+			defer r.wg.Done()
+			r.registrationLoop()
+		}()
+		r.lggr.Info("started")
+		return nil
+	})
+}
+
+func (r *Client) Close() error {
+	return r.StopOnce(r.lggr.Name(), func() error {
+		close(r.stopCh)
+		r.wg.Wait()
+		r.lggr.Info("closed")
+		return nil
+	})
+}
+
+func (r *Client) RegisterWorkflow(workflowID string, request []byte) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.lggr.Infow("register workflow called", "capabilityId", r.capInfo.ID, "donId", r.capDonInfo.ID, "workflowID", workflowID)
+	regState, ok := r.registeredWorkflows[workflowID]
+	if !ok {
+		regState = &clientRegistration{
+			registrationRequest: request,
+		}
+		r.registeredWorkflows[workflowID] = regState
+	} else {
+		regState.registrationRequest = request
+		r.lggr.Warnw("re-registering workflow", "capabilityId", r.capInfo.ID, "donId", r.capDonInfo.ID, "workflowID", workflowID)
+	}
+
+	return nil
+}
+
+func (r *Client) UnregisterWorkflow(workflowID string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.lggr.Infow("unregister workflow called", "capabilityId", r.capInfo.ID, "donId", r.capDonInfo.ID, "workflowID", workflowID)
+	delete(r.registeredWorkflows, workflowID)
+	// Registrations will quickly expire on all remote nodes so it is currently considered unnecessary to send
+	// unregister messages to the nodes
+}
+
+func (r *Client) registrationLoop() {
```

Comment:
this is pretty much the same as the current registrationLoop() in triggerSubscriber, right?